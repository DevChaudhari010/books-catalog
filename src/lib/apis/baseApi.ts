import axios from "axios";

const host = process.env.API_SERVER;

const axiosInstance = axios.create({
  baseURL: host + "/api/v1",
});

export default axiosInstance;
