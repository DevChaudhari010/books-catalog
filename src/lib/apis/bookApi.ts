import { IBook } from "@types";
import baseApi from "./baseApi";

export type GetBooksResponse = {
  books: IBook[] | null;
  error: boolean;
};

const getBooks: () => Promise<GetBooksResponse> = async () => {
  try {
    const response = await baseApi({ method: "get", url: "books" });
    return { books: response.data, error: false };
  } catch (error) {
    return { error: true, books: null };
  }
};

export type GetBookDetailsResponse = {
  book: IBook | null;
  error: boolean;
};

const getBookDetails: (id: string) => Promise<GetBookDetailsResponse> = async (
  bookId
) => {
  try {
    const response = await baseApi({
      method: "get",
      url: `books/${bookId}`,
    });
    return { book: response.data, error: false };
  } catch (error) {
    return { error: true, book: null };
  }
};

export default { getBooks, getBookDetails };
