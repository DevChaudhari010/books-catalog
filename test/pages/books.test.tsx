import { render } from "@testing-library/react";
import "@testing-library/jest-dom/extend-expect";
import Books, { getStaticProps } from "@pages/books";
import { IBook } from "@types";
import { bookApi } from "@lib/apis";

jest.mock("@lib/apis");

const books: IBook[] = [
  { _id: "1", title: "Test 1" },
  { _id: "2", title: "Test 2" },
];

describe("Books page", () => {
  test("matches snapshot for books", () => {
    const { asFragment } = render(<Books books={books} error={false} />, {});
    expect(asFragment()).toMatchSnapshot();
  });
  test("matches snapshot for error", () => {
    const { asFragment } = render(<Books books={null} error={true} />, {});
    expect(asFragment()).toMatchSnapshot();
  });
  test("Check book title exists", () => {
    const { getByText } = render(<Books books={books} error={false} />, {});
    expect(getByText("Test 1")).toBeInTheDocument();
    expect(getByText("Test 2")).toBeInTheDocument();
  });
  test("Check error messgae exists on Error", () => {
    const { getByText } = render(<Books books={null} error={true} />, {});
    expect(
      getByText("Failed to load books. Please try again after sometime")
    ).toBeInTheDocument();
  });
  test("Check getStaticProps", async () => {
    const response = { books, error: false };
    //@ts-ignore
    bookApi.getBooks.mockImplementation(() => Promise.resolve(response));
    const staticProps = await getStaticProps({});
    expect(bookApi.getBooks).toBeCalled();
    expect(staticProps).toEqual(
      expect.objectContaining({
        props: response,
      })
    );
  });
});
