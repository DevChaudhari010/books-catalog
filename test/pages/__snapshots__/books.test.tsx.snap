// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Books page matches snapshot for books 1`] = `
<DocumentFragment>
  <div
    class="container"
  >
    <h2>
      To know more about book click on Book
    </h2>
    <div
      class="grid"
    >
      <a
        class="card"
        href="/books/1"
      >
        <h2>
          Test 1
        </h2>
      </a>
      <a
        class="card"
        href="/books/2"
      >
        <h2>
          Test 2
        </h2>
      </a>
    </div>
  </div>
</DocumentFragment>
`;

exports[`Books page matches snapshot for error 1`] = `
<DocumentFragment>
  <div>
    Failed to load books. Please try again after sometime
  </div>
</DocumentFragment>
`;
